src/main/java/com/limechain/trie/TrieAccessor.java

```diff
@@ -0,0 +1,227 @@
+                    .orElse(null);
+            case NodeHandle<NodeData> handle -> handle.getParent();
+        };
+
+        return Optional.ofNullable(parent)
+                .map(NodeHandle::getFullKey)
+                .map(parentKey -> prefixIndexInParent(parentKey, prefix))
+                .flatMap(parent::getChild)
+                .map(c -> lexicographicDelete(c, limit))
+                .orElse(new DeleteByPrefixResult(0, true));
+
+    }
+
+    private Nibble prefixIndexInParent(Nibbles parent, Nibbles prefix) {
+        for (int i = 0; i < prefix.size(); i++) {
+            if(prefix.get(i).equals(parent.get(i))) {
+                return parent.get(i);
+            }
+        }
+        return null;
+    }
+
+    private DeleteByPrefixResult lexicographicDelete(NodeHandle<NodeData> node, Long limit) {
+        loadChildren(node.getFullKey());
+        int deleted = 0;
+        for (Nibble nibble : Nibbles.ALL) {
+            if (limit != null && limit <= deleted) {
+                return new DeleteByPrefixResult(deleted, false);
+            }
+            NodeHandle<NodeData> child = node.getChild(nibble).orElse(null);
+            if (child != null) {
+                DeleteByPrefixResult childResult = lexicographicDelete(child, limit == null ? null : limit - deleted);
+                deleted += childResult.deleted();
+                if (!childResult.all()) {
+                    return new DeleteByPrefixResult(deleted, false);
+                }
+            }
+        }
+        if (partialTrie.clearNodeValue(node.getNodeIndex())) {
+            deleted++;
+        }
+        return new DeleteByPrefixResult(deleted, true);
+    }
+
+    public void persistAll() {
+        for (ChildTrieAccessor value : loadedChildTries.values()) value.persistAll();
+        loadedChildTries.clear();
+
+        if (partialTrie.isEmpty()) return;
+        lastRoot = getMerkleRoot(StateVersion.V1);
+        trieStorage.insertTrieStorage(partialTrie, StateVersion.V1);
+    }
+
+    @Override
+    public Optional<Nibbles> getNextKey(Nibbles key) {
+        loadPathToKey(key);
+        //TODO: optimize using partial trie
+        return Optional.ofNullable(trieStorage.getNextKeyByMerkleValue(lastRoot, key.toString()))
+                .map(k -> Nibbles.fromBytes(k.getBytes()));
+    }
+
+    @Override
+    public void startTransaction() {
+        throw new UnsupportedOperationException(TRANSACTIONS_NOT_SUPPORTED);
+
+    }
+
+    @Override
+    public void rollbackTransaction() {
+        throw new UnsupportedOperationException(TRANSACTIONS_NOT_SUPPORTED);
+
+    }
+
+    @Override
+    public void commitTransaction() {
+        throw new UnsupportedOperationException(TRANSACTIONS_NOT_SUPPORTED);
+
+    }
+
+    @Override
+    public void closeConnection() {
+        throw new UnsupportedOperationException(TRANSACTIONS_NOT_SUPPORTED);
+    }
+
+    public byte[] getMerkleRoot(StateVersion version) {
```

Comment:
I really don't like the state version being a public argument here, since this binds this class to always be used in a combination with a `Runtime` instance (which knows the state version). For now this concern might not require a given refactor, since this is a broader topic (hybrid state is also not handled, so we'll have to think about it); I'm just leaving this comment as a mental note somewhere.